server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStream.java

```diff
@@ -99,4 +110,38 @@ public final int available() throws IOException {
         return stream == null ? 0 : stream.available();
     }
 
+    @Override
+    public boolean markSupported() {
+        return true;
+    }
+
+    @Override
+    public void mark(int readLimit) {
+        // We ignore readLimit since openSlice() can re-open previous InputStreams, and we can skip as many bytes as we'd like.
+        // According to JDK documentation, marking a closed InputStream should have no effect.
+        if (closed == false) {
+            if (initialized && nextSlice > 0) { // nextSlice > 0 guards the case it is initialized and numSlices is 0.
```

Comment:
How it possible for `initialized = true` and `nextSlice = 0` unless there are concurrent invocation on nextStream and mark? But this class seems to assume single threaded usage?

Reply:
Sorry, maybe my comment misled you to think of concurrency.

There's indeed a single-threaded edge case where it can happen: if you create a SlicedInputStream with 0 slices (effectively an empty InputStream), and try to read, then the SlicedInputStream is initialized and nextSlice = 0. That's why I had added the nextSlice>0 guard here.

However, I reworked the code to make it better: I added an explicit numSlices > 0 guard to both mark/reset to make them ineffective. And added a test `testMarkResetZeroSlices` that covers this edge case.